### PR TITLE
[Linter] Redundant boolean expressions

### PR DESCRIPTION
**Pull request by dzungdinh94**
# Description
This rules detects and warns about unnecessary boolean operation such as x || true .

# Run the linter
```
cargo test move_check_testsuit
```

# Testing
File testing: redundant_boolean.move

### CODE REVIEW

File: external-crates/move/crates/move-compiler/src/sui_mode/linters/custom_rules/redundant_boolean_expressions.rs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    ) -> Self::Context<'a> {
+        Context { env }
+    }
+}
+
+impl TypingVisitorContext for Context<'_> {
+    fn visit_exp_custom(&mut self, exp: &mut T::Exp) -> bool {
+        if let UnannotatedExp_::BinopExp(e1, op, _, e2) = &exp.exp.value {
+            // Check if the operation is a logical OR
+            if let BinOp_::Or = &op.value {
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/sui_mode/linters/custom_rules/redundant_boolean_expressions.rs:51]: We could also do this for `&&`, `==`, or `!=`, albeit with different rules

File: external-crates/move/crates/move-compiler/src/sui_mode/linters/custom_rules/redundant_boolean_expressions.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                match (&e1.exp.value, &e2.exp.value) {
+                    (UnannotatedExp_::Value(bool), UnannotatedExp_::Copy { var, .. })
+                    | (UnannotatedExp_::Copy { var, .. }, UnannotatedExp_::Value(bool)) => {
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/sui_mode/linters/custom_rules/redundant_boolean_expressions.rs:54]: This should work with any expression, no?

**Review by tnowacki — CHANGES_REQUESTED** [review]
Let's extend this to any boolean expression. That is, there is no real reason to use `!`, `||`, `&&`, `==`, or `!=` with boolean literals. Though the diagnostic/lint message will have to consider the case to give good help

File: external-crates/move/crates/move-compiler/src/linters/mod.rs [hunk 3]
```diff
@@ -17,16 +20,33 @@ pub enum LintLevel {
 
 pub const ALLOW_ATTR_CATEGORY: &str = "lint";
 pub const LINT_WARNING_PREFIX: &str = "Lint ";
+pub const REDUNDANT_BOOLEAN_EXP_FILTER_NAME: &str = "redundant_boolean_expressions";
+
+pub const LINTER_DEFAULT_DIAG_CODE: u8 = 1;
+
+pub enum LinterDiagCategory {
+    RedundantBooleanExp,
+}
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/linters/mod.rs:29]: See comment on #16488, we should make sure we have a good conceptual split between category and code

File: external-crates/move/crates/move-compiler/tests/custom_rules/redundant_boolean.move [hunk 4]
```diff
@@ -0,0 +1,16 @@
+module 0x42::M {
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/custom_rules/redundant_boolean.move:1]: Let's move this to `tests/linter`

File: external-crates/move/crates/move-compiler/tests/custom_rules/redundant_boolean.move [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    // Function with non-redundant boolean expression
+    public fun non_redundant_expression(y: bool): bool {
+        y && true // Should not trigger a warning
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/custom_rules/redundant_boolean.move:4]: Why shouldn't this trigger a warning?
```
false && true ==> false
true && true ==> true
```

Comment (tnowacki) [reply external-crates/move/crates/move-compiler/tests/custom_rules/redundant_boolean.move:4]: Similarly 
```
x && false === false
```

File: external-crates/move/crates/move-compiler/src/linters/redundant_boolean_expressions.rs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        Context { env }
+    }
+}
+
+impl TypingVisitorContext for Context<'_> {
+    fn visit_exp_custom(&mut self, exp: &mut T::Exp) -> bool {
+        if let UnannotatedExp_::BinopExp(e1, op, _, e2) = &exp.exp.value {
+            // Check if the operation is a logical OR
+            if let BinOp_::Or = &op.value {
+                match (&e1.exp.value, &e2.exp.value) {
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/linters/redundant_boolean_expressions.rs:52]: We should really handle any and all redundant boolean expressions, and merge this with 
#16488

File: external-crates/move/crates/move-compiler/src/linters/redundant_boolean_expressions.rs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        _program_info: &'a TypingProgramInfo,
+        _program: &T::Program_,
+    ) -> Self::Context<'a> {
+        Context { env }
+    }
+}
+
+impl TypingVisitorContext for Context<'_> {
+    fn visit_exp_custom(&mut self, exp: &mut T::Exp) -> bool {
+        if let UnannotatedExp_::BinopExp(e1, op, _, e2) = &exp.exp.value {
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/linters/redundant_boolean_expressions.rs:49]: See comments on #16488 for let-else

**Review by tnowacki — CHANGES_REQUESTED** [review]

File: external-crates/move/crates/move-compiler/src/linters/redundant_boolean_expressions.rs [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+            // Check if the operation is a logical OR
+            if let BinOp_::Or = &op.value {
+                match (&e1.exp.value, &e2.exp.value) {
+                    (UnannotatedExp_::Value(bool), UnannotatedExp_::Copy { var, .. })
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/linters/redundant_boolean_expressions.rs:53]: This is still overly constrained. I would consider all `BinOp` where the LHS/RHS is `true` or `false`. This only considers `Or` with `copy var`.
Consider `&&`, `||`, `==`, and `!=` with `true` and `false`. All of those cases are redundant in some manner.

**Review by tnowacki — CHANGES_REQUESTED** [review]

File: external-crates/move/crates/move-compiler/src/linters/redundant_boolean_expressions.rs (hunk 8 above)
Comment (dzungdinh94) [reply external-crates/move/crates/move-compiler/src/linters/redundant_boolean_expressions.rs]: merged to #16488

**Pull request closed without merging**